File: maths/typescript/test/utils/readTestData.ts
import type { StableState } from '@/stable/data';
import type { WeightedState } from '@/weighted/data';
import type { LiquidityBootstrappingState } from '@/liquidityBootstrapping';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { FixedPriceLBPState } from '@/fixedPriceLBP';
import { ReClammState } from '@/reClamm';
import { type HookState } from '../../src/hooks/types';
import { mapHookState } from './mapHookState';
import { Vault } from '@/vault/vault';

type HookData = {
    address: string;
    type: string;
    enableHookAdjustedAmounts: boolean;
    shouldCallAfterSwap: boolean;
    shouldCallBeforeSwap: boolean;
    shouldCallAfterInitialize: boolean;
    shouldCallBeforeInitialize: boolean;
    shouldCallAfterAddLiquidity: boolean;
    shouldCallBeforeAddLiquidity: boolean;
    shouldCallAfterRemoveLiquidity: boolean;
    shouldCallBeforeRemoveLiquidity: boolean;
    shouldCallComputeDynamicSwapFee: boolean;
    dynamicData?: Record<string, string>;
};

type PoolBase = {
    chainId: number;
    blockNumber: number;
    poolAddress: string;
    hookType?: string;
    hook?: HookState;
};

type WeightedPool = PoolBase & WeightedState;

type StablePool = PoolBase & StableState;





type LiquidityBootstrappingPool = PoolBase & LiquidityBootstrappingState;


type FixedPriceLBPPool = PoolBase & FixedPriceLBPState;

type ReClammPool = PoolBase & ReClammState;

type SupportedPools =
    | WeightedPool
    | StablePool
    | LiquidityBootstrappingPool
    | FixedPriceLBPPool
    | ReClammPool;

type PoolsMap = Map<string, SupportedPools>;

type Swap = {
    swapKind: number;
    amountRaw: bigint;
    outputRaw: bigint;
    tokenIn: string;
    tokenOut: string;
    test: string;
};

type Add = {
    kind: number;
    inputAmountsRaw: bigint[];
    bptOutRaw: bigint;
    test: string;
};

type Remove = {
    kind: number;
    amountsOutRaw: bigint[];
    bptInRaw: bigint;
    test: string;
};

type TestData = {
    swaps: Swap[];
    adds: Add[];
    pools: PoolsMap;
    removes: Remove[];
};

// Reads all json test files and parses to relevant swap/pool bigint format
export function readTestData(directoryPath: string): TestData {
    const pools: PoolsMap = new Map<string, SupportedPools>();
    const swaps: Swap[] = [];
    const adds: Add[] = [];
    const removes: Remove[] = [];
    const testData: TestData = {
        swaps,
        adds,
        pools,
        removes,
    };

    // Resolve the directory path relative to the current file's directory
    const absoluteDirectoryPath = path.resolve(__dirname, directoryPath);

    // Read all files in the directory
    const files = fs.readdirSync(absoluteDirectoryPath);

    // Iterate over each file
    for (const file of files) {
        // Check if the file ends with .json
        if (file.endsWith('.json')) {
            // Read the file content
            const fileContent = fs.readFileSync(
                path.join(absoluteDirectoryPath, file),
                'utf-8',
            );

            // Parse the JSON content
            try {
                const jsonData = JSON.parse(fileContent);
                // A pool type this package has no maths for yet (maths may land one language at a time).
                if (!new Vault().supportsPoolType(jsonData.pool.poolType)) {
                    console.warn(`${file}: pool type ${jsonData.pool.poolType} not implemented in TypeScript, skipped`);
                    continue;
                }
                if (jsonData.swaps)
                    swaps.push(
                        ...jsonData.swaps.map((swap) => ({
                            ...swap,
                            swapKind: Number(swap.swapKind),
                            amountRaw: BigInt(swap.amountRaw),
                            outputRaw: BigInt(swap.outputRaw),
                            test: file,
                        })),
                    );
                if (jsonData.adds)
                    adds.push(
                        ...jsonData.adds.map((add) => ({
                            ...add,
                            kind: add.kind === 'Unbalanced' ? 0 : 1,
                            inputAmountsRaw: add.inputAmountsRaw.map((a) =>
                                BigInt(a),
                            ),
                            bptOutRaw: BigInt(add.bptOutRaw),
                            test: file,
                        })),
                    );
                if (jsonData.removes)
                    removes.push(
                        ...jsonData.removes.map((remove) => ({
                            ...remove,
                            kind: mapRemoveKind(remove.kind),
                            amountsOutRaw: remove.amountsOutRaw.map((a) =>
                                BigInt(a),
                            ),
                            bptInRaw: BigInt(remove.bptInRaw),
                            test: file,
                        })),
                    );

                pools.set(file, mapPool(jsonData.pool));
            } catch (error) {
                console.error(`Error parsing JSON file ${file}:`, error);
            }
        }
    }

    return testData;
}

/**
 * The `pool` block of a test-data file → the pool state the Vault expects. Generic: every decimal string becomes a
 * bigint (recursively, so hook data is covered too); addresses, booleans and JSON numbers (token indices) pass
 * through; the hook block, if any, becomes the hook state. A new pool type therefore needs no changes here — the
 * generator writes the fields its maths reads (`getReClammPoolDynamicData()`, `weights`, `amp`, …) and the pool
 * class receives them under the same names.
 */
function mapPool(pool: Record<string, unknown>): SupportedPools {
    const state = toBigints(pool) as Record<string, unknown>;
    // The generator writes the pool's setting; older hand-made files may lack it.
    state.supportsUnbalancedLiquidity = pool.supportsUnbalancedLiquidity ?? true;
    // Time-dependent pools read the timestamp the queries were made at; files without one are timeless.
    state.currentTimestamp = state.currentTimestamp ?? BigInt(Date.now());

    if (pool.hook) {
        const hookState = mapHookState(pool.hook as HookData, {
            tokens: pool.tokens as string[],
            amp: state.amp as bigint | undefined,
        });
        state.hookType = hookState.hookType;
        state.hook = hookState;
    }
    return state as unknown as SupportedPools;
}

// Decimal strings → bigint, recursively. Matches what maths/check's runners do with the same data.
function toBigints(x: unknown): unknown {
    if (Array.isArray(x)) return x.map(toBigints);
    if (x && typeof x === 'object')
        return Object.fromEntries(
            Object.entries(x).map(([k, v]) => [k, toBigints(v)]),
        );
    if (typeof x === 'string' && /^[0-9]+$/.test(x)) return BigInt(x);
    return x;
}

function mapRemoveKind(kind: string): number {
    if (kind === 'Proportional') return 0;
    else if (kind === 'SingleTokenExactIn') return 1;
    else if (kind === 'SingleTokenExactOut') return 2;
    else throw new Error(`Unsupported RemoveKind: ${kind}`);
}

File: maths/rust/tests/utils/read_test_data.rs
//! Reads every test-data file in maths/testData into swaps / adds / removes plus the pool (and hook) state each
//! one was recorded against. Pool states are built by `bush_maths::pools::state_from_json`, so a new pool type
//! needs nothing here.

use alloy_primitives::U256;
use bush_maths::common::json::PoolJson;
use bush_maths::common::types::PoolState;
use bush_maths::hooks::types::HookState;
use bush_maths::pools::state_from_json;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::Path;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HookData {
    pub address: String,
    #[serde(rename = "type")]
    pub hook_type: String,
    #[serde(rename = "dynamicData")]
    pub dynamic_data: serde_json::Value,
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct Swap {
    pub swap_kind: u8,
    pub amount_raw: U256,
    pub output_raw: U256,
    pub token_in: String,
    pub token_out: String,
    pub test: String,
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct Add {
    pub kind: u8,
    pub input_amounts_raw: Vec<U256>,
    pub bpt_out_raw: U256,
    pub test: String,
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct Remove {
    pub kind: u8,
    pub amounts_out_raw: Vec<U256>,
    pub bpt_in_raw: U256,
    pub test: String,
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct TestData {
    pub swaps: Vec<Swap>,
    pub adds: Vec<Add>,
    pub removes: Vec<Remove>,
    /// Keyed by file name.
    pub pools: HashMap<String, PoolState>,
    pub hook_states: HashMap<String, HookState>,
}

#[derive(Deserialize)]
struct RawTestFile {
    swaps: Option<Vec<RawSwap>>,
    adds: Option<Vec<RawAdd>>,
    removes: Option<Vec<RawRemove>>,
    pool: serde_json::Value,
}

#[derive(Deserialize)]
struct RawSwap {
    #[serde(rename = "swapKind")]
    swap_kind: u8,
    #[serde(rename = "amountRaw")]
    amount_raw: String,
    #[serde(rename = "outputRaw")]
    output_raw: String,
    #[serde(rename = "tokenIn")]
    token_in: String,
    #[serde(rename = "tokenOut")]
    token_out: String,
}

#[derive(Deserialize)]
struct RawAdd {
    kind: String,
    #[serde(rename = "inputAmountsRaw")]
    input_amounts_raw: Vec<String>,
    #[serde(rename = "bptOutRaw")]
    bpt_out_raw: String,
}

#[derive(Deserialize)]
struct RawRemove {
    kind: String,
    #[serde(rename = "amountsOutRaw")]
    amounts_out_raw: Vec<String>,
    #[serde(rename = "bptInRaw")]
    bpt_in_raw: String,
}

type BoxError = Box<dyn std::error::Error>;

fn parse_all(xs: &[String]) -> Result<Vec<U256>, BoxError> {
    xs.iter().map(|x| Ok(x.parse::<U256>()?)).collect()
}

pub fn read_test_data() -> Result<TestData, BoxError> {
    let mut data = TestData {
        swaps: Vec::new(),
        adds: Vec::new(),
        removes: Vec::new(),
        pools: HashMap::new(),
        hook_states: HashMap::new(),
    };

    let dir = Path::new(env!("CARGO_MANIFEST_DIR")).parent().unwrap().join("testData");
    let mut files: Vec<_> = fs::read_dir(dir)?
        .filter_map(|e| e.ok().map(|e| e.path()))
        .filter(|p| p.extension().is_some_and(|x| x == "json"))
        .collect();
    files.sort();

    for path in files {
        let test = path.file_name().unwrap().to_string_lossy().to_string();
        let file: RawTestFile = serde_json::from_str(&fs::read_to_string(&path)?)
            .map_err(|e| format!("{test}: {e}"))?;

        // A pool type this crate has no maths for yet (maths may land one language at a time).
        let pool = PoolJson(file.pool.clone());
        let pool_type = pool.s("poolType")?;
        if state_from_json(&pool, None).is_err_and(|e| e.starts_with("Unsupported pool type")) {
            eprintln!("{test}: pool type {pool_type} not implemented in Rust, skipped");
            continue;
        }

        for s in file.swaps.unwrap_or_default() {
            data.swaps.push(Swap {
                swap_kind: s.swap_kind,
                amount_raw: s.amount_raw.parse()?,
                output_raw: s.output_raw.parse()?,
                token_in: s.token_in,
                token_out: s.token_out,
                test: test.clone(),
            });
        }
        for a in file.adds.unwrap_or_default() {
            data.adds.push(Add {
                kind: if a.kind == "Unbalanced" { 0 } else { 1 },
                input_amounts_raw: parse_all(&a.input_amounts_raw)?,
                bpt_out_raw: a.bpt_out_raw.parse()?,
                test: test.clone(),
            });
        }
        for r in file.removes.unwrap_or_default() {
            data.removes.push(Remove {
                kind: map_remove_kind(&r.kind),
                amounts_out_raw: parse_all(&r.amounts_out_raw)?,
                bpt_in_raw: r.bpt_in_raw.parse()?,
                test: test.clone(),
            });
        }

        // The hook, if any, is mapped first: the pool state carries the hook type the Vault dispatches on.
        let hook_state = match file.pool.get("hook").filter(|h| !h.is_null()) {
            Some(hook) => {
                let hook_data: HookData = serde_json::from_value(hook.clone())?;
                let pool_data = crate::utils::PoolData {
                    tokens: pool.strings("tokens")?,
                    amp: pool.u("amp").ok(),
                    weights: pool.arr("weights").ok(),
                    swap_fee: pool.u("swapFee")?,
                };
                Some(crate::utils::map_hook_state(&hook_data, &pool_data)?)
            }
            None => None,
        };
        let hook_type = hook_state.as_ref().map(|h| match h {
            HookState::StableSurge(_) => "StableSurge".to_string(),
            HookState::Custom(state) => state.hook_type.clone(),
        });

        data.pools.insert(test.clone(), state_from_json(&pool, hook_type).map_err(|e| format!("{test}: {e}"))?);
        if let Some(h) = hook_state {
            data.hook_states.insert(test, h);
        }
    }

    Ok(data)
}

fn map_remove_kind(kind: &str) -> u8 {
    match kind {
        "Proportional" => 0,
        "SingleTokenExactIn" => 1,
        "SingleTokenExactOut" => 2,
        _ => panic!("Unsupported RemoveKind: {}", kind),
    }
}
